src/beamlime/applications/_nexus_helpers.py

```diff
@@ -1,155 +1,365 @@
-) -> Optional[Mapping]:
-    '''Returns the branch or leaf associated with `path`, or None if not found'''
+def find_nexus_structure(structure: Mapping, path: Sequence[Optional[str]]) -> Mapping:
+    """Returns the branch or leaf associated with `path`, or None if not found"""
     if len(path) == 0:
         return structure
     head, *tail = path
-    for child in structure['children']:
+    for child in structure["children"]:
         if head == _node_name(child):
             return find_nexus_structure(child, tail)
+    raise KeyError(f"Path {path} not found in the nexus structure.")
+
+
+def find_parent(structure: Mapping, path: Sequence[Optional[str]]) -> DataModuleParent:
+    # TODO: We can use typeguard here later.
+    return find_nexus_structure(structure, path[:-1])
 
 
-def path_to_message_group(structure: Mapping, message: Mapping) -> str:
-    '''Finds the group where the message should be merged'''
-    kind, content = message
+def find_ev44_matching_paths(
+    structure: Mapping, data_piece: Mapping
+) -> Iterable[NexusPath]:
+    source_name = data_piece["source_name"]
     for path, node in iter_nexus_structure(structure):
         if (
-            node.get('module') == kind
-            and node.get('config', {}).get('source') == content['source_name']
+            node.get("module") == "ev44"
+            and node.get("config", {}).get("source") == source_name
         ):
-            return path
+            yield path
 
 
-def merge_message_into_store(
-    store: Mapping[Tuple[Optional[str], ...], Mapping],
+def _merge_message_into_data_module_store(
+    *,
+    data_module_store: DataModuleStore,
     structure: Mapping,
-    message: Tuple[str, Mapping],
+    data_piece: Mapping,
+    path_matching_func: Callable[[Mapping, Mapping], Iterable[tuple[str | None, ...]]],
+    data_field_initialize_func: Callable[[DataModuleParent], None],
+    merge_func: Callable[[DataModuleParent, Mapping], None],
+) -> None:
+    """Bridge function to merge a message into the store.
+
+    This bridge was needed since different modules
+    have different ways to match the paths and merge the data.
+
+    Parameters
+    ----------
+    data_module_store:
+        The data module store that holds the data.
+    structure:
+        The nexus structure.
+    data_piece:
+        The content of the message.
+    path_matching_func:
+        A function that returns the paths that match the message.
+    data_field_initialize_func:
+        A function that initializes the datasets in the module.
+        *Initialize is done only when the relevant message arrives.*
+        It is because we should distinguish between empty dataset and
+        unexpectedly-not-receiving data.
+    merge_func:
+        A function that merges the message into the store.
+
+    Side Effects
+    ------------
+    The ``data_module_store`` is updated with the merged data.
+
+    """
+    for path in path_matching_func(structure, data_piece):
```

Comment:
I don't understand this. Should the message be merged in multiple places?

Reply:
Yes... unfortunately....

Reply:
This is news to me, can you give an example just so I understand how it's supposed to work?